### ISSUE TITLE
Conntrack table not flushed for UDP hostPort when pod deleted

### ISSUE THREAD

**Issue by kinghrothgar** [kubernetes/kubernetes#95258]
**What happened**:
Restart daemonset with UDP hostPort. All UDP traffic to new ds pods from existing pods fails. Flushing conntrack table fixes it. I saw this issue with 1.14 and it seems to be the exact same issue as here https://github.com/kubernetes/kubernetes/issues/66651. It was fixed in 1.15 and now I'm seeing it again in 1.16.

**What you expected to happen**:
hostPorts to work across Pod restarts

**How to reproduce it (as minimally and precisely as possible)**:
Restart daemonset with UDP hostPort. UDP traffic to new pod does not work.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.9", GitCommit:"a17149e1a189050796ced469dbd78d380f2ed5ef", GitTreeState:"clean", BuildDate:"2020-04-16T11:44:51Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.13-gke.401", GitCommit:"eb94c181eea5290e9da1238db02cfef263542f5f", GitTreeState:"clean", BuildDate:"2020-09-09T00:57:35Z", GoVersion:"go1.13.9b4", Compiler:"gc", Platform:"linux/amd64"}
- Cloud provider or hardware configuration:
GKE
- OS (e.g: `cat /etc/os-release`):
COS
- Kernel (e.g. `uname -a`):
Linux 4.19.112+ #1 SMP Fri Sep 4 12:00:04 PDT 2020 x86_64 Intel(R) Xeon(R) CPU @ 2.30GHz GenuineIntel GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
Vanilla GKE
- Others:

**Comment by aojea** [comment]
@kinghrothgar just to confirm, you are referring to container hostPort only, right? no Service NodePorts.

```
kind: Pod
apiVersion: v1
metadata:
  name: udp
  labels:
    app: udp-app
spec:
  containers:
  - name: container
    image: reg/udpdaemon:1.0
    ports:
    - name: udp
      hostPort: 55555
      containerPort: 80      
      protocol: UDP
```

**Comment by kinghrothgar** [comment]
Correct, there is no service in the mix here.

**Comment by kinghrothgar** [comment]
I have not been able to reproduce this issue in an isolated cluster using `nc`. The current behavior I am seeing is occurring between Datadog agents and code clients. I will continue to debugging and update as I find more information.

**Comment by aojea** [comment]
hmm, this is tricky, because the hostport on containers is managed differently depending on the runtime. The dockershim host port manager seems to clear stale udp entries at first sight

https://github.com/kubernetes/kubernetes/blob/19064f47389429f8360e1d6684bef3d770db9368/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L180-L187

@thockin how are hostports managed in `Vanilla GCE`?

**Comment by kinghrothgar** [comment]
Is there anything else I can do to help debug this?

**Comment by aojea** [comment]
ideally, steps to repro and cluster configuration (cni, cri, ...)

**Comment by kinghrothgar** [comment]
Alright, my current theory is maybe there's a race condition in the clearing of conntrack entries and creation of the iptables rules pointing at the new IP. Here's the setup and steps I took.

Pod IP info:
```
server 10.16.32.54
client  10.16.32.53
```

Output of conntrack while in working state:
```
HOSTNAME ~ # conntrack -L -p udp --dport 8125
udp      17 18 src=10.16.32.53 dst=10.142.0.25 sport=35926 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=35926 mark=0 use=1
udp      17 15 src=10.16.32.53 dst=10.142.0.25 sport=54303 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=54303 mark=0 use=1
udp      17 25 src=10.16.32.53 dst=10.142.0.25 sport=50295 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=50295 mark=0 use=1
udp      17 29 src=10.16.32.53 dst=10.142.0.25 sport=56531 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=56531 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 4 flow entries have been shown.
```

I then delete the server pod and the replacement's IP is `10.16.32.57`

It is now in the broken state. This is the output for the conntracks:
```
HOSTNAME ~ # conntrack -L -p udp --dport 8125
udp      17 17 src=10.16.32.53 dst=10.142.0.25 sport=35926 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.53 sport=8125 dport=35926 mark=0 use=1
udp      17 22 src=10.16.32.53 dst=10.142.0.25 sport=50295 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=50295 mark=0 use=1
udp      17 24 src=10.16.32.55 dst=10.142.0.25 sport=45306 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.55 sport=8125 dport=45306 mark=0 use=1
udp      17 29 src=10.16.32.53 dst=10.142.0.25 sport=56531 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=56531 mark=0 use=2
conntrack v1.4.4 (conntrack-tools): 4 flow entries have been shown.
```
As you can see, some entries have been updated to the new correct `src=10.16.32.57` while others still have the old `src=10.16.32.54`. I then delete the UDP entries:
```
HOSTNAME ~ # conntrack -D -p udp --dport 8125
udp      17 2 src=10.16.32.53 dst=10.142.0.25 sport=35926 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.53 sport=8125 dport=35926 mark=0 use=1
udp      17 25 src=10.16.32.23 dst=10.142.0.25 sport=51310 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.23 sport=8125 dport=51310 mark=0 use=1
udp      17 24 src=10.16.32.53 dst=10.142.0.25 sport=50295 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=50295 mark=0 use=1
udp      17 25 src=10.16.32.22 dst=10.142.0.25 sport=47319 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.22 sport=8125 dport=47319 mark=0 use=1
udp      17 29 src=10.16.32.55 dst=10.142.0.25 sport=45306 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.55 sport=8125 dport=45306 mark=0 use=1
udp      17 29 src=10.16.32.53 dst=10.142.0.25 sport=56531 dport=8125 [UNREPLIED] src=10.16.32.54 dst=10.16.32.53 sport=8125 dport=56531 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 6 flow entries have been deleted.
```
This fixes the issue. Communication resumes between client and server. This is the output of conntrack now:
```
HOSTNAME ~ # conntrack -L -p udp --dport 8125
udp      17 27 src=10.16.32.53 dst=10.142.0.25 sport=35926 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.53 sport=8125 dport=35926 mark=0 use=1
udp      17 27 src=10.16.32.53 dst=10.142.0.25 sport=50295 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.53 sport=8125 dport=50295 mark=0 use=1
udp      17 12 src=10.16.32.53 dst=10.142.0.25 sport=44676 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.53 sport=8125 dport=44676 mark=0 use=1
udp      17 24 src=10.16.32.55 dst=10.142.0.25 sport=45306 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.55 sport=8125 dport=45306 mark=0 use=1
udp      17 29 src=10.16.32.53 dst=10.142.0.25 sport=56531 dport=8125 [UNREPLIED] src=10.16.32.57 dst=10.16.32.53 sport=8125 dport=56531 mark=0 use=1
conntrack v1.4.4 (conntrack-tools): 5 flow entries have been shown.
```

**Comment by kinghrothgar** [comment]
I trying to debug this further and logs that should be getting printed out are not and I'm not sure what I'm doing wrong. I set my log level to 4 and restarted kubelet. I verified that kubelet correctly printed out `FLAG: --v="4"`. However when I delete the server pod, I see neither of these logs:
https://github.com/kubernetes/kubernetes/blob/v1.16.13/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L249
https://github.com/kubernetes/kubernetes/blob/v1.16.13/pkg/util/iptables/iptables.go#L397
Following the code path for the code you posted above these should be getting hit. Does this mean that GKE is a different code path than @aojea posted? Any feedback from @thockin on GKE hostport?

**Comment by kinghrothgar** [comment]
Also, to add more info, I do not have network policy enabled (so I shouldn't be using calico) but I do have VPC native enabled. I am seeing these logs which implies to me I am using a CNI plugin though:
```
I1029 20:00:39.466786 1976800 plugins.go:407] Calling network plugin cni to set up pod "datadog-bl64c_monitoring"
l1029 20:00:39.469308 1976800 cni.go:361] Adding monitoring_datadog-bl64c/35ef94a421060fe1358c9fc9284253f4643febd5c3bed9e653d735a84021c53a to network loopback/cni-loopback netns "/proc/1983519/ns/net"
```
I believe I have found the CNI binary and running `gke --version` prints out: `CNI ptp plugin v0.8.5-gke.1`

**Comment by kinghrothgar** [comment]
Alright, I don't believe this is a race condition, it's a matter of the old entries just aren't removed. I deleted the pod with this command running: `conntrack -E -p udp --dport 8125` which monitors all changes to the conntrack table. This was the output:
```
[2020-10-30 00:21:57.853650] [DESTROY] udp      17 src=10.16.32.23 dst=10.142.0.25 sport=50994 dport=8125 [UNREPLIED] src=10.16.32.75 dst=10.16.32.23 sport=8125 dport=50994
[2020-10-30 00:21:59.925657] [DESTROY] udp      17 src=10.16.32.22 dst=10.142.0.25 sport=34789 dport=8125 [UNREPLIED] src=10.16.32.75 dst=10.16.32.22 sport=8125 dport=34789
[2020-10-30 00:22:02.698183]     [NEW] udp      17 30 src=10.16.32.23 dst=10.142.0.25 sport=50994 dport=8125 [UNREPLIED] src=10.16.32.76 dst=10.16.32.23 sport=8125 dport=50994
[2020-10-30 00:22:02.700768]     [NEW] udp      17 30 src=10.16.32.22 dst=10.142.0.25 sport=34789 dport=8125 [UNREPLIED] src=10.16.32.76 dst=10.16.32.22 sport=8125 dport=34789
```
All the NEW events contain the correct `src=10.16.32.76` which is the new pod's IP. However as shown in previous messages conntrack entries with the IP remain and are not cleaned up and never do get cleaned up.

**Comment by aojea** [comment]
> I trying to debug this further and logs that should be getting printed out are not and I'm not sure what I'm doing wrong. I set my log level to 4 and restarted kubelet.

I 've added verbosity for conntrack operations recently https://github.com/kubernetes/kubernetes/pull/95694, it will be available  in 1.20 :shrug: 

Anways, try to find the cni config, you should see the configuration options in the kubelet logs, it prints it when it is starting

```
I1104 19:46:17.370452   10532 flags.go:59] FLAG: --cni-bin-dir="/home/kubernetes/bin"
I1104 19:46:17.370459   10532 flags.go:59] FLAG: --cni-cache-dir="/var/lib/cni/cache"
I1104 19:46:17.370466   10532 flags.go:59] FLAG: --cni-conf-dir="/etc/cni/net.d"
```
in this case, you should have a file in `/etc/cni/net.d` and we can try to figure out if you are using the CNI portmap plugin or the hostport_manager

**Comment by thockin** [comment]
Is this running dockershim on the node or containerd?  Could be that containerd is never clearing conntrack?

**Comment by kinghrothgar** [comment]
From searching around I believe that GKE uses docker by default. I know when I log into a node I use docker to interact with the pods running on the node, so I assume that means I'm using dockershim.

I believe the issue is that GKE is using netd as it's CNI which I assume is blocked by this issue that has seen no movement https://github.com/containernetworking/plugins/issues/123. Thus the fix / code path  merged here https://github.com/kubernetes/kubernetes/pull/59286 never gets run. Does it seem like I am understanding this correctly?

**Comment by kinghrothgar** [comment]
More details that might be helpful:
```
/home/kubernetes/bin/kubelet
 --v=2
 --cloud-provider=gce
 --experimental-check-node-capabilities-before-mount=true
 --experimental-mounter-path=/home/kubernetes/containerized_mounter/mounter
 --cert-dir=/var/lib/kubelet/pki/
 --cni-bin-dir=/home/kubernetes/bin
 --kubeconfig=/var/lib/kubelet/kubeconfig
 --image-pull-progress-deadline=5m
 --experimental-kernel-memcg-notification=true
 --max-pods=35
 --non-masquerade-cidr=0.0.0.0/0
 --network-plugin=cni
 --register-with-taints=dedicated=taskpool:NoSchedule,cloud.google.com/gke-preemptible=true:NoSchedule
 --node-labels=cloud.google.com/gke-netd-ready=true,cloud.google.com/gke-nodepool=taskpool-preemptible-staging,cloud.google.com/gke-os-distribution=cos,cloud.google.com/gke-preemptible=true,dedicated=taskpool,iam.gke.io/gke-metadata-server-enabled=true,size=e2-standard-4
 --volume-plugin-dir=/home/kubernetes/flexvolume
 --node-status-max-images=25
 --registry-qps=10
 --registry-burst=20
 --config /home/kubernetes/kubelet-config.yaml
 --pod-sysctls=net.core.somaxconn=1024,net.ipv4.conf.all.accept_redirects=0,net.ipv4.conf.all.forwarding=1,net.ipv4.conf.all.route_localnet=1,net.ipv4.conf.default.forwarding=1,net.ipv4.ip_forward=1,net.ipv4.tcp_fin_timeout=60,net.ipv4.tcp_keepalive_intvl=75,net.ipv4.tcp_keepalive_probes=9,net.ipv4.tcp_keepalive_time=7200,net.ipv4.tcp_rmem=4096 87380 6291456,net.ipv4.tcp_syn_retries=6,net.ipv4.tcp_tw_reuse=0,net.ipv4.tcp_wmem=4096 16384 4194304,net.ipv4.udp_rmem_min=4096,net.ipv4.udp_wmem_min=4096,net.ipv6.conf.default.accept_ra=0,net.netfilter.nf_conntrack_generic_timeout=600,net.netfilter.nf_conntrack_tcp_timeout_close_wait=3600,net.netfilter.nf_conntrack_tcp_timeout_established=86400
```

And I varified that this is the configmap that is live in my cluster configuring netd: https://github.com/GoogleCloudPlatform/netd/blob/master/netd.yaml#L67

**Comment by AnishShah** [comment]
@kinghrothgar This is definitely an issue with portmap CNI plugin. GKE CNI is chained with portmap CNI to create those DNAT iptable rules to forward packet from hostport to the container. Unfortunately, portmap is not flushing conntrack entries when the container is deleted.

**Comment by AnishShah** [comment]
We will need a similar fix like kubernetes/kubernetes#59286 but for portmap plugin. But I'm not sure if that's the correct way to fix it.  I have commented on containernetworking/plugins#123 to start the discussion.

**Comment by aojea** [comment]
The fix in the in portmap plugin was merged https://github.com/containernetworking/plugins/pull/553

@kinghrothgar can you try to use the latest version of the plugin and provide feedback?

**Comment by kinghrothgar** [comment]
I will try to this coming week

**Comment by bowei** [comment]
@vbannai

**Comment by kinghrothgar** [comment]
It does not look like I can test this easily. The version of portmap GKE 1.16 is currently using is different enough that placing the new portmap binary breaks basic functions:

```
Failed create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "02fd0d13b86456e7c041fc8a2c71cfb8be4654fc0e6d07310956690525a6a3be" network for pod "datadog-dmnx5": networkPlugin cni failed to set up pod "datadog-dmnx5_monitoring" network: netplugin failed with no error message
```

If anyone has suggestions on how I can easily test this I will.

**Comment by kinghrothgar** [comment]
As @AnishShah tested this as stated here https://github.com/containernetworking/plugins/pull/553#issuecomment-737287696, I will assume this issue is indeed fixed. Once a new release version of plugins is cut, I will reach out to Google to see when a release of GKE can be put out that contains the fix.

**Comment by DanTulovsky** [comment]
Have you had a chance to find out when GKE might support this fix?

Thank you
Dan

**Comment by bowei** [comment]
@vbannai -- do we have a timeline on what CNI binaries are used in which release?

**Comment by aexvir** [comment]
@DanTulovsky @bowei got confirmed by gcp support that the fix is included with `1.21.4-gke.2300`
and they can't confirm that the fix will get backported to any other version

message from gcp support:
> Looking further into this I see this fix was included with the CNI release 0.9.0 [2], currently on GKE only versions starting after the release “1.21.4-gke.2300” will include CNI versions with this fix, looking at the GKE regular channel release notes I see that currently GKE “1.21.5-gke.1302” is available which would include this fix.
